### PR TITLE
progress: clean up Progress and add Throbber classes

### PR DESCRIPTION
**Pull request by bmr-cymru**
Resolves: #664
Resolves: #665
Resolves: #666
Resolves: #667
Resolves: #668
Resolves: #669
Resolves: #670
Resolves: #671
Resolves: #672
Resolves: #673
Resolves: #674

## Summary by CodeRabbit

* **New Features**
  * Terminal-aware animated throbbers (full, ASCII fallback and silent) added and exposed via the public API; progress factory can now select throbbers. New cancel() hook for progress flows.

* **Bug Fixes**
  * Guarded stream flush to avoid broken-pipe errors; improved cursor handling, initial render behaviour and end-of-operation cleanup.

* **Tests**
  * Expanded coverage for progress/throbber lifecycles, factory selection, terminal edge cases, error paths and rendering markers.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: snapm/_progress.py [hunk 1]
```diff
@@ -714,6 +749,231 @@ def _do_end(self, _message: Optional[str] = None):
...
+        self._frame_index: int = 0
+        self._interval_us: int = round((1.0 / self.fps) * _USECS_PER_SEC)
+        self._last: Optional[datetime] = None
+
+    def start(self):
+        """
+        Begin a throbber run with the specified ``total``.
+
+        :param total: The total number of expected throbber items.
+        :type total: ``int``
```
[the remark this answers is not in the document]
Comment (bmr-cymru) [reply snapm/_progress.py]: Fixed in commit ea6052d.

File: tests/test_progress.py [hunk 2]
```diff
@@ -280,12 +337,51 @@ def test_lifecycle_no_clear(self):
...
+        self.assertIn("Halfway", output)
+
+        # 3. Canecl while in progres
+        mock_tc.term_stream.truncate(0)
+        mock_tc.term_stream.seek(0)
+        p.cancel("Quit!")
+        output = mock_tc.term_stream.getvalue()
+        self.assertIn("<SHOW_CURSOR>", output)
+        self.assertIn("Quit!", output)
+
```
[the remark this answers is not in the document]
Comment (bmr-cymru) [reply tests/test_progress.py]: Fixed in commit 48ef26d.

File: snapm/_progress.py [hunk 3]
```diff
@@ -714,6 +753,230 @@ def _do_end(self, _message: Optional[str] = None):
...
+        Maintain liveness for this throbber and output frame if required.
+        """
+        self._check_started()
+        now = datetime.now()
+        if (now - self._last).total_seconds() * _USECS_PER_SEC >= self._interval_us:
+            self._do_throb()
+            _flush_with_broken_pipe_guard(self.stream)
+            self._last = now
+            self._frame_index = (self._frame_index + 1) % self.nr_frames
+            self.first_update = False
```
[the remark this answers is not in the document]
Comment (bmr-cymru) [reply snapm/_progress.py]: Fixed in commit 48ef26d.
